raft/fsm.cc

```diff
@@ -45,11 +45,14 @@ fsm::fsm(server_id id, term_t current_term, server_id voted_for, log log,
         failure_detector& failure_detector, fsm_config config) :
         fsm(id, current_term, voted_for, std::move(log), index_t{0}, failure_detector, config) {}
 
-future<semaphore_units<>> fsm::wait_for_memory_permit(seastar::abort_source* as, size_t size) {
+future<fsm::memory_permit> fsm::wait_for_memory_permit(seastar::abort_source* as, size_t size) {
     check_is_leader();
 
-    auto& sm = *leader_state().log_limiter_semaphore;
-    return as ? get_units(sm, size, *as) : get_units(sm, size);
+    memory_permit mp;
+    mp.sem = leader_state().log_limiter_semaphore;
+    auto& sm = *mp.sem;
+    mp.units = co_await (as ? get_units(sm, size, *as) : get_units(sm, size));
```

Comment:
The fact that get_units may return unit to a broken semaphore is just a bug in its implementation. It should not be workarounded in the user code.

Reply:
According to @avikivity it's by design, and it's the caller responsibility to take care of the semaphore liveliness.

I'd also sleep better if outstanding units are tracked by the semaphore and cover special cases like moving the semaphore (updating any backpointers if needed), breaking it (that should gracefully release the units), and destroying it (that should assert there are no outstanding units).

The reason this wasn't done was that the gained safety would come on the expense of common case performance.

Reply:
Where that design was written and discussed? No need to explain every bug in the code which is not trivial to fix by a design. 

> The reason this wasn't done was that the gained safety would come on the expense of common case performance.

It is not only about safety. Why require all units to be freed when destroying a semaphore. Here we have a use case where we want to destroy a semaphore without releasing all units.

Reply:
Because the units are returned, by default, when destroyed, using a pointer to the semaphore. Therefore moving the semaphore or destroying it while there are outstanding units is unsafe.

Reply:
> Because the units are returned, by default, when destroyed, using a pointer to the semaphore. Therefore moving the semaphore or destroying it while there are outstanding units is unsafe.

The semaphore is not moved here. Here we get units back after the semaphore is already destroyed. Just bad API. But before your changes to the semaphore it worked since releasing the units did not touch the semaphore. Why was it changed?

Reply:
The more I look at it the more I think that the seastar part needs to be reverted. It breaks the semaphore in the debug mode to catch some illegal moves. If you could do it without breaking the semaphore fine, but no need to break working code to do it. And those moves are probably caught by memory sanitizer anyway. Did you see the assert trigger even once after introducing it?

Reply:
> > Because the units are returned, by default, when destroyed, using a pointer to the semaphore. Therefore moving the semaphore or destroying it while there are outstanding units is unsafe.
> 
> The semaphore is not moved here. Here we get units back after the semaphore is already destroyed. Just bad API. But before your changes to the semaphore it worked since releasing the units did not touch the semaphore. Why was it changed?

It was changed to catch move of the semaphore while it has outstanding units before use-after-move happened since the sanitizer wouldn't catch that, unless the moved-from semaphore is also destroyed.

> The more I look at it the more I think that the seastar part needs to be reverted. It breaks the semaphore in the debug mode to catch some illegal moves. If you could do it without breaking the semaphore fine, but no need to break working code to do it. And those moves are probably caught by memory sanitizer anyway.

Unfortunately not in all cases, see above.

> Did you see the assert trigger even once after introducing it?

No, not until the issue discussed here that popped up with the change to update seastar.

Reply:
> > > Because the units are returned, by default, when destroyed, using a pointer to the semaphore. Therefore moving the semaphore or destroying it while there are outstanding units is unsafe.
> > 
> > 
> > The semaphore is not moved here. Here we get units back after the semaphore is already destroyed. Just bad API. But before your changes to the semaphore it worked since releasing the units did not touch the semaphore. Why was it changed?
> 
> It was changed to catch move of the semaphore while it has outstanding units before use-after-move happened since the sanitizer wouldn't catch that, unless the moved-from semaphore is also destroyed.

But this is likely the case. Otherwise why move from it?

> 
> > The more I look at it the more I think that the seastar part needs to be reverted. It breaks the semaphore in the debug mode to catch some illegal moves. If you could do it without breaking the semaphore fine, but no need to break working code to do it. And those moves are probably caught by memory sanitizer anyway.
> 
> Unfortunately not in all cases, see above.
> 
> > Did you see the assert trigger even once after introducing it?
> 
> No, not until the issue discussed here that popped up with the change to update seastar.

The issue here is no something the new code catches. On the contrary it is a use case it breaks and we get nothing in return. Not a single bug found.

Reply:
Can't we just take all the available units in destroy() in a high priority manner? That would ensure the semaphore being destroyed has no outstanding units. There can be no new units taken by others, we only need to wait for oustanding units.

Reply:
what do you mean by `destroy()`?
`~leader()` is synchronous so we can't simply wait on `log_limiter_semaphore.wait(whatever)` there.

Destroying the `leader` does "no new units taken by others" like you mentioned above
and "wait for oustanding units" is achieved in this patch by extending the semaphore lifetime, so I think it achieves the goal in a simple way.

Reply:
~leader() is synchronous, but we can and are actually already destroying the semaphore in server::abort()